src/runtime_src/driver/xclng/drm/xocl/subdev/nifd.c

```diff
@@ -0,0 +1,680 @@
+
+    for (i = 0; i < num_breakpoints; ++i)
+    {
+        frame_address = (*arg);
+        ++arg;
+        frame_offset = (*arg);
+        ++arg;
+        constraint = (*arg);
+        ++arg;
+
+        add_breakpoint_data(1, frame_address, frame_offset, constraint);
+    }
+
+    breakpoint_condition = (*arg);
+
+    write_nifd_register(breakpoint_condition, NIFD_BREAKPOINT_CONDITION);
+
+    return 0; // Success
+}
+
+static long add_breakpoints(void __user *arg)
+{
+    // Format of user data: [numBreakpoints][frameAddress][frameOffset][constraint]...[condition]
+    unsigned int num_breakpoints;
+    unsigned int total_data_size;
+    unsigned int *kernel_memory;
+    long result;
+
+    if (copy_from_user(&num_breakpoints, arg, sizeof(unsigned int)))
+        return -EFAULT;
+
+    total_data_size = (num_breakpoints * 3 + 1 + 1) * sizeof(unsigned int);
+    kernel_memory = (unsigned int *)(vmalloc(total_data_size));
+    if (!kernel_memory)
+        return -ENOMEM;
+
+    if (copy_from_user(kernel_memory, arg, total_data_size))
+    {
+        vfree(kernel_memory);
+        return -EFAULT;
+    }
+
+    result = add_breakpoints_core(kernel_memory);
+    vfree(kernel_memory);
+
+    if (result)
+        return result;
+    return 0;
+}
+
+static long remove_breakpoints(void)
+{
+    unsigned int clock_status = (read_nifd_status() & 0x3);
+    stop_controlled_clock();
+    clear_configuration_memory(0);
+    write_nifd_register(0x1, NIFD_CLEAR);
+    restart_controlled_clock(clock_status);
+
+    return 0;
+}
+
+static long check_status(void __user *arg)
+{
+    unsigned int status = read_nifd_status();
+    if (copy_to_user(arg, &status, sizeof(unsigned int)))
+    {
+        return -EFAULT;
+    }
+    return 0; // Success
+}
+
+static long nifd_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
+{
+    // struct xocl_nifd *nifd = filp->private_data;
+    long status = 0;
+
+    void __user *data = (void __user *)(arg);
+
+    switch (cmd)
+    {
+    case NIFD_STOP_CONTROLLED_CLOCK:
+        status = stop_controlled_clock();
+        break;
+    case NIFD_START_CONTROLLED_CLOCK:
+        status = start_controlled_clock(data);
+        break;
+    case NIFD_SWITCH_ICAP_TO_NIFD:
+        status = switch_icap_to_nifd();
+        break;
+    case NIFD_SWITCH_ICAP_TO_PR:
+        status = switch_icap_to_pr();
+        break;
+    case NIFD_READBACK_VARIABLE:
+        status = readback_variable(data);
+        break;
+    case NIFD_SWITCH_CLOCK_MODE:
+        status = switch_clock_mode(data);
+        break;
+    case NIFD_ADD_BREAKPOINTS:
+        status = add_breakpoints(data);
+        break;
+    case NIFD_REMOVE_BREAKPOINTS:
+        status = remove_breakpoints();
+        break;
+    case NIFD_CHECK_STATUS:
+        status = check_status(data);
+        break;
+    default:
+        status = -ENOIOCTLCMD;
+        break;
+    }
+
+    return status;
+}
+
+static int char_open(struct inode *inode, struct file *file)

```

Comment:
Do you want to follow what is done in xvc's open/close routines to avoid race that Lizhi has fixed?

Reply:
I changed it in one of the following commits. @houlz0507 please take a look and see if that is the correct way to do it. Thanks!

Reply:
looks good to me